### PR TITLE
IWF-364: Fix conditional complete on messages of internalChannels from state APIs

### PR DESCRIPTION
**Pull request by longquanzheng**
### Checklist
- [x] Code compiles correctly
- [ ] Tests for the changes have been added
- [x] All tests passing
- [x] **This PR change is backwards-compatible**
- [ ] **This PR CONTAINS a (planned) breaking change** (it is not backwards compatible)

### Related Issue
Closes #issue_number

### CODE REVIEW

File: service/interpreter/globalVersioner.go [hunk 1]
```diff
@@ -65,20 +65,22 @@ func (p *GlobalVersioner) IsAfterVersionOfNoIwfGlobalVersionSearchAttribute() bo
...
 func (p *GlobalVersioner) UpsertGlobalVersionSearchAttribute() error {
-	if p.OmitVersionMarker {
-		// the search attribute is already set when starting the workflow
-		return nil
-	}
-	// TODO this bug in Cadence SDK may cause concurrent writes
-	// https://github.com/uber-go/cadence-client/issues/1198
-	if p.workflowProvider.GetBackendType() != service.BackendTypeCadence {
+	if p.IsUsingGlobalVersionSearchAttribute() &&
+		p.workflowProvider.GetBackendType() != service.BackendTypeCadence {
```
Comment (longquanzheng) [inline service/interpreter/globalVersioner.go:80]: this is just some minor cleanup that is not related to this ticket

File: service/interpreter/signalReceiver.go [hunk 2]
```diff
@@ -239,15 +239,16 @@ func (sr *SignalReceiver) DumpReceived(channelNames []string) map[string][]*iwfi
...
+// workflow task.
 // There are two cases this is needed:
 // 1. ContinueAsNew:
 // retrieve signals that after signal handler threads are stopped,
 // so that the signals can be carried over to next run by continueAsNew.
 // This includes both regular user signals and system signals
 // 2. Conditional close/complete workflow on signal/internal channel:
 // retrieve all signal/internal channel messages before checking the signal/internal channels
-func (sr *SignalReceiver) DrainAllUnreceivedSignals(ctx UnifiedContext) {
+func (sr *SignalReceiver) DrainAllReceivedButUnprocessedSignals(ctx UnifiedContext) {
```
Comment (longquanzheng) [inline service/interpreter/signalReceiver.go:251]: Rename the method to make it more clear(hopefully)

File: service/interpreter/workflowImpl.go [hunk 3]
```diff
@@ -487,6 +486,23 @@ func checkClosingWorkflow(
...
+		// Just yield, by waiting on an empty lambda, nothing else.
+		// It will let other workflow threads/coroutines to run.
+		// This will drain the messages published from state APIs.
+		// NOTE that this is extremely tricky in Cadence/Temporal programming model.
+		// Read more: https://stackoverflow.com/questions/71356668/how-does-multi-threading-works-in-cadence-temporal-workflow
+		//https://docs.temporal.io/encyclopedia/go-sdk-multithreading
+		return provider.Await(ctx, func() bool {
+			return true
+		})
+	}
```
Comment (lwolczynski) [inline service/interpreter/workflowImpl.go:502]: I'm gonna trust this works as expected. Having tests would be beneficial and make it easier to understand and verify. Hope we can cover multi-threading soon so I can get a better grasp on it.

**Review by lwolczynski — APPROVED** [review]

**Comment by lwolczynski** [comment]
Closes #289
